app/models/guest_device.rb

```diff
@@ -10,4 +11,7 @@ class GuestDevice < ApplicationRecord
 
   has_one :network, :foreign_key => "device_id", :dependent => :destroy, :inverse_of => :guest_device
   has_many :miq_scsi_targets, :dependent => :destroy
+
+  has_many :firmwares, :dependent => :destroy
+  has_many :child_devices, -> { where(:parent_device_id => ids) }, :foreign_key => "parent_device_id", :class_name => "GuestDevice", :dependent => :destroy
```

Comment:
That sounds like a bug in the refresh, honestly.

Can you update the PR with the changes you have in place, and maybe it will shed some light on what's going on.

Also, with the inventory refresher code... are you running what's on master?  Or, do you have local changes to the inventory collection?  It would help to see that, too.